src/ucs/sys/event_set.c

```diff
@@ -0,0 +1,150 @@
+
+ucs_status_t ucs_event_set_create(ucs_sys_event_set_t **event_set_p)
+{
+    ucs_sys_event_set_t *event_set;
+    ucs_status_t status;
+
+    event_set = ucs_malloc(sizeof(ucs_sys_event_set_t), "ucs_sys_event_set");
+    if(event_set == NULL){
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+
+    /* Create epoll set the thread will wait on */
+    event_set->epfd = epoll_create(1);
+    if (event_set->epfd < 0) {
+        ucs_error("epoll_create() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_free;
+    }
+
+    status = UCS_OK;
+    *event_set_p = event_set;
+    goto out_create;
+
+err_free:
+    ucs_free(event_set);
+
+err:
+
+out_create:
+    return status;
+}
+
+ucs_status_t ucs_event_set_add(ucs_sys_event_set_t *event_set, int event_fd,
+                               int events, void *udata)
+{
+    struct epoll_event event;
+    ucs_status_t status;
+    int ret;
+
+    memset(&event, 0, sizeof(event));
+    event.events = events;
+    event.data.fd = event_fd;
+    event.data.ptr = udata;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_ADD, event_fd, &event);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, ADD, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        status = UCS_ERR_IO_ERROR;
+        goto err_add_event;
+    }
+
+    return UCS_OK;
+
+err_add_event:
+
+    return status;
+
+}
+
+ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+{
+    ucs_status_t status;
+    int ret;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_DEL, event_fd, NULL);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, DEL, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        status = UCS_ERR_IO_ERROR;
+        goto err_del_event;
+    }
+
+    return UCS_OK;
+
+err_del_event:
+
+    return status;
+}
+
+//
+// event 
+//  epoll:  epoll_event.data.ptr 
+//  kqueue: kevent.udata
+//
+ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
+                                ucs_status_t (*event_set_handler)(int nready,
+                                                                  void *event))
+{
+    int nready;
+    ucs_status_t status;
+//    struct epoll_event events[UCS_ASYNC_EPOLL_MAX_EVENTS];
+    struct epoll_event *event;
+
+    event  = ucs_malloc(sizeof(struct epoll_event) * UCS_ASYNC_EPOLL_MAX_EVENTS,
+                        "epoll_event");
+    if( event == NULL ) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+
+    nready = epoll_wait(event_set->epfd, event, UCS_ASYNC_EPOLL_MAX_EVENTS,
+                        timeout_ms);
+
+    if ((nready < 0) && (errno != EINTR)) {
+        ucs_fatal("epoll_wait() failed: %m");
+        goto err_free_event;
+    }
+    ucs_trace_async("epoll_wait(epfd=%d, timeout=%d) returned %d",
+                    event_set->epfd, timeout_ms, nready);
+
+    status = event_set_handler(nready,event);
```

Comment:
How to separate the following part?
https://github.com/openucx/ucx/blob/master/src/ucs/async/thread.c#L101-L118

I think this part `ucs/async/thread.c` specific code.


Just simply write in `ucs/sys/event_set.c` with `#if HAVE_EPOLL`?

```C
        /* Check ready files */
        if (nready > 0) {
            for (i = 0; i < nready; ++i) {
                fd = events[i].data.fd;

                /* Check wakeup pipe */
                if (fd == ucs_async_pipe_rfd(&thread->wakeup)) {
                    ucs_trace_async("progress thread woken up");
                    ucs_async_pipe_drain(&thread->wakeup);
                    continue;
                }

                status = ucs_async_dispatch_handlers(&fd, 1);
                if (status == UCS_ERR_NO_PROGRESS) {
                    is_missed = 1;
                }
            }
        }

        /* Check timers */
        curr_time = ucs_get_time();
        if (curr_time - last_time > timer_interval) {
            status = ucs_async_dispatch_timerq(&thread->timerq, curr_time);
            if (status == UCS_ERR_NO_PROGRESS) {
                 is_missed = 1;
            }

            last_time = curr_time;
        }
```

Reply:
the code above should remain in async/thread.c and NOT move to event.c